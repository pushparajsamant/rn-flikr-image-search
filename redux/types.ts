export interface ImageType {
  id: string;
  owner: string;
  secret: string;
  server: string;
  farm: number;
  title: string;
  ispublic: number;
  isfriend: number;
  isfamily: number;
}

export interface PhotosState {
  page: number;
  pages: number;
  perpage: number;
  total: number;
  photo: ImageType[];
  error: string | undefined;
  search: string;
  oldSearches: string[];
}
export interface PhotosResponse {
  photos: PhotosState;
  stat: string;
}
