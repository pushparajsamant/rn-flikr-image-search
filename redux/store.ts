import {combineReducers, configureStore} from '@reduxjs/toolkit';
import SearchReducer from './reducers/SearchReducer';
import {useDispatch} from 'react-redux';
const rootReducer = combineReducers({
  search: SearchReducer,
});
const store = configureStore({
  reducer: rootReducer,
  //middleware: getDefaultMiddleware => getDefaultMiddleware().concat(logger),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export default store;
