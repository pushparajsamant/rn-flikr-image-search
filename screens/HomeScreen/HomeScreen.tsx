import React from 'react';
import {searchFlickrAPI} from '../../redux/reducers/SearchReducer';
import {NavigationProp} from '@react-navigation/native';
import {RootState, useAppDispatch} from '../../redux/store';
import {FlatList, SafeAreaView, StyleSheet, Text, View} from 'react-native';
import {useSelector} from 'react-redux';
import Tile from '../../components/Tile/Tile';
import SearchInput from '../../components/SearchInput/SearchInput';
import {StackNavigatorParams} from '../../navigation/MainNavigation';

interface Props {
  navigation?: NavigationProp<StackNavigatorParams>;
}
const HomeScreen: React.FC<Props> = ({navigation}) => {
  const dispatch = useAppDispatch();
  const searchState = useSelector((state: RootState) => state.search);
  function openSearchPage() {
    //Navigate to search page to show previous searches
    navigation && navigation.navigate('SearchScreen');
  }

  return (
    <SafeAreaView>
      <SearchInput
        onFocus={() => openSearchPage()}
        value={searchState.search}
      />
      {searchState.error !== '' && searchState.error !== undefined ? (
        <View style={styles.error}>
          <Text>{'Error'}</Text>
        </View>
      ) : (
        <FlatList
          data={searchState.photo}
          //This makes the flat list 2 columned
          numColumns={2}
          keyExtractor={(item, index) => item.id + index}
          renderItem={({item}) => <Tile item={item} />}
          onEndReachedThreshold={0.5}
          //Calling reducer function here to trigger infinite loading
          onEndReached={async () => {
            await searchFlickrAPI(
              searchState.search,
              dispatch,
              searchState.page + 1,
            );
          }}
        />
      )}
    </SafeAreaView>
  );
};
export default HomeScreen;
const styles = StyleSheet.create({
  error: {
    justifyContent: 'center',
    alignItems: 'center',
  },
});
