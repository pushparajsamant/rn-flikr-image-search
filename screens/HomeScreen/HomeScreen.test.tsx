// components/HomeScreen.test.js
import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import HomeScreen from './HomeScreen';
import {PhotosState} from '../../redux/types';

const mockStore = configureStore([]);
jest.mock('react-native-device-info', () => {
  return {
    hasNotch: () => true,
  };
});
jest.mock('@fortawesome/react-native-fontawesome', () => ({
  FontAwesomeIcon: '',
}));

test('HomeScreen renders correctly with data from Redux', () => {
  const dataFromRedux: PhotosState = {
    page: 7,
    pages: 556,
    perpage: 20,
    total: 11118,
    search: 'kohli',
    oldSearches: ['kohli'],
    error: '',
    photo: [
      {
        id: '52708477344',
        owner: '197759494@N02',
        secret: 'e883e24554',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708707438',
        owner: '197759494@N02',
        secret: '0e034643f9',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52707700162',
        owner: '197759494@N02',
        secret: '2772dd9ea6',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708477294',
        owner: '197759494@N02',
        secret: '252cbca428',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708635975',
        owner: '197759494@N02',
        secret: 'cc17ab1e00',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708477239',
        owner: '197759494@N02',
        secret: '8a4c427f3b',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708635885',
        owner: '197759494@N02',
        secret: '9fb75886fe',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708707258',
        owner: '197759494@N02',
        secret: '7f50bcf173',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708221451',
        owner: '197759494@N02',
        secret: '254355b453',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708221406',
        owner: '197759494@N02',
        secret: 'fde6d3c379',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708635745',
        owner: '197759494@N02',
        secret: '5eaa3d1741',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708477069',
        owner: '197759494@N02',
        secret: '06ce4163fb',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708477039',
        owner: '197759494@N02',
        secret: '8c95d685d2',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708635670',
        owner: '197759494@N02',
        secret: '8630771f26',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52707699647',
        owner: '197759494@N02',
        secret: '3355106abf',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708635485',
        owner: '197759494@N02',
        secret: 'bee7f0fd19',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
      {
        id: '52708221041',
        owner: '197759494@N02',
        secret: '65c386332b',
        server: '65535',
        farm: 66,
        title: 'Rahul Kohli',
        ispublic: 1,
        isfriend: 0,
        isfamily: 0,
      },
    ],
  };
  const store = mockStore({search: dataFromRedux});

  const tree = renderer
    .create(
      <Provider store={store}>
        <HomeScreen />
      </Provider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test('HomeScreen renders correctly with error', () => {
  const store = mockStore({someData: null});

  const tree = renderer.create().toJSON();

  //expect(tree).toMatchSnapshot();
});
