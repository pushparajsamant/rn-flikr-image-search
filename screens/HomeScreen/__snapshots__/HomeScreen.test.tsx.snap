// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders correctly with data from Redux 1`] = `
<RCTSafeAreaView>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 5,
        "marginVertical": 21.516129032258068,
        "paddingHorizontal": 21.428571428571427,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#CAF0F8",
          "borderColor": "#CAF0F8",
          "borderRadius": 32.14285714285714,
          "borderWidth": 1,
          "flexDirection": "row",
          "flexGrow": 3,
          "height": 107.58064516129033,
          "paddingHorizontal": 21.428571428571427,
        }
      }
    >
      <
        color="#25C0FF"
        icon={
          {
            "icon": [
              512,
              512,
              [
                128269,
                "search",
              ],
              "f002",
              "M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z",
            ],
            "iconName": "magnifying-glass",
            "prefix": "fas",
          }
        }
        size={22}
      />
      <TextInput
        autoCorrect={false}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Search"
        style={
          {
            "flex": 1,
            "fontFamily": "Inter",
            "fontSize": 27.906976744186046,
            "height": "100%",
            "lineHeight": 31.3953488372093,
            "marginLeft": 12.857142857142858,
            "paddingHorizontal": 34.285714285714285,
          }
        }
        value=""
      />
    </View>
  </View>
  <RCTScrollView
    data={
      [
        {
          "farm": 66,
          "id": "52708477344",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "e883e24554",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708707438",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "0e034643f9",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52707700162",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "2772dd9ea6",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708477294",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "252cbca428",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708635975",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "cc17ab1e00",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708477239",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "8a4c427f3b",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708635885",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "9fb75886fe",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708707258",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "7f50bcf173",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708221451",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "254355b453",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708221406",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "fde6d3c379",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708635745",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "5eaa3d1741",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708477069",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "06ce4163fb",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708477039",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "8c95d685d2",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708635670",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "8630771f26",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52707699647",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "3355106abf",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708635485",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "bee7f0fd19",
          "server": "65535",
          "title": "Rahul Kohli",
        },
        {
          "farm": 66,
          "id": "52708221041",
          "isfamily": 0,
          "isfriend": 0,
          "ispublic": 1,
          "owner": "197759494@N02",
          "secret": "65c386332b",
          "server": "65535",
          "title": "Rahul Kohli",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708477344_e883e24554.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708707438_0e034643f9.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52707700162_2772dd9ea6.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708477294_252cbca428.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708635975_cc17ab1e00.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708477239_8a4c427f3b.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708635885_9fb75886fe.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708707258_7f50bcf173.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708221451_254355b453.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708221406_fde6d3c379.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708635745_5eaa3d1741.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708477069_06ce4163fb.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708477039_8c95d685d2.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708635670_8630771f26.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52707699647_3355106abf.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708635485_bee7f0fd19.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://farm66.static.flickr.com/65535/52708221041_65c386332b.jpg",
              }
            }
            style={
              {
                "height": 100,
                "margin": 8.571428571428571,
                "width": "48%",
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
